Dataset Sampling Backend: `astartes` (`chemprop` v2) + updates to workflows, env, pyproject, setup

JacksonBurns opened the pull request:
See https://github.com/chemprop/chemprop/pull/432#issuecomment-1658908548 and the surrounding discussion.

JacksonBurns commented:
TODO: add the repeated smiles sampler

hwpang commented on chemprop/data/utils.py:
```diff
@@ -1,5 +1,6 @@
 from collections import OrderedDict, defaultdict
 import csv
+import copy
```
I think changes in this file should be within the `chemprop/v2/data/utils.py`?

hwpang commented on tests/unit/featurizers/test_atom.py:
```diff
@@ -5,7 +5,7 @@
 from rdkit.Chem.rdchem import HybridizationType
 
-from chemprop.v2.featurizers.multihot import AtomFeaturizer
+from chemprop.v2.featurizers import AtomFeaturizer
```
I think the changes in this file should be in `chemprop/v2/tests/unit/featurizers/test_atom.py` and I think @am2145 addresses most of the changes in this commit in his PR https://github.com/chemprop/chemprop/pull/463. It might make sense to rebase from his once his is merged.

hwpang commented on chemprop/v2/data/datasets.py:
```diff
@@ -36,6 +36,7 @@ def __len__(self) -> int:
     @cached_property
     def _Y(self) -> np.ndarray:
         """the raw targets of the dataset"""
+        print(self)
```
Can you remove the print

kevingreenman commented:
@JacksonBurns thanks for getting this ready! Could you write a few bullet points explaining your reasons for changes in the test workflows, environment, pyproject, and setup files? I think we've discussed most of this in meetings, but I think it might help me and other developers remember/understand the reasons for the changes unrelated to splitting

JacksonBurns commented:
> test workflows

 - now runs on v2/dev, which we will remove in the future
 - runs tests on all platforms
 - fixed the conda environment resolution
 - reordered test and linting, temporarily

> environment

 - removed the dependencies on pip, since they are also on conda and it is better to be all in once place
 - specified new python verison

> pyproject

 - move version attribute into pyproject
 - added the new dependency for splitting molecules

> setup files

 - removed setup.cfg, since pyproject.toml contains all of the information

JacksonBurns commented:
@kevingreenman unit tests are actually running now (fixed a small typo) and the explanations are above. Let me know any other comments!

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
-    k: int = 5,
-    fold: int = 0,
-):
+    sizes: Tuple[float, float, float] = (0.8, 0.1, 0.1),
```
`sizes: tuple[...`
The current convention in v2 has been to use lowercase types (tuple, list), which was added in Python 3.9
Not a must change, but could be good.

JacksonBurns replied:
The outer types have to be capitalized since they are not built-ins, so I think I will leave this as is

KnathanM replied:
After further discussion, we decided to keep it as lowercase tuple. Not backwards compatible with older python versions but consistent with the rest of the v2 code.

KnathanM replied:
Bumping that Tuple should be replaced with the built in version and the import removed.

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
-):
+    sizes: Tuple[float, float, float] = (0.8, 0.1, 0.1),
+    key_molecule_index: int = 0,
+    seed: int = 0,
```
If the seed defaults to 0, will this give the same "random" split each time Chemprop is run? It could be better to default seed to 'None'.

kevingreenman replied:
The seed should be set to 0 by default to follow chemprop's existing practice of being reproducible without needing to explicitly set a seed. People who want different splits each time will have to specify different seeds / use cross-validation.

JacksonBurns replied:
This is also practice in `astartes`.

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
+    key_molecule_index: int = 0,
+    seed: int = 0,
+    num_folds: int = 1,
+    args: TrainArgs = None,
```
The 'TrainArgs' class is no more. We'll have to find another way to get the args needed into this function.

JacksonBurns replied:
41d28da addresses this. Unforunately train args was used for three of the splitting approaches (different ways of loading splits from files) so we will indeed need to add that back. I think we should only add back one way to do it tough. v1 had three approaches which did slightly different things and were all poorly documented.

KnathanM replied:
This will be added back in another PR, after more discussion with the team about the best way to do it.

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
-        for idxs in scaffold2idxs.values():
-            if len(idxs) > n_val / 2 or len(idxs) > n_test / 2:
-                big_index_sets.append(idxs)
+    random = Random(seed)
```
Is this object 'random' used? If so, I feel it could be moved closer to where it is used.

JacksonBurns replied:
this is (was) still used by some of the samplers i had not yet moved to `astartes`. I have made the changes locally and will push them later - just finishing porting the repeated smiles sampler

JacksonBurns replied:
Done

KnathanM commented on environment.yml:
```diff
@@ -19,6 +20,5 @@ dependencies:
-    - typed-argument-parser>=1.6.1
-    - git+https://github.com/bp-kelley/descriptastorus
+  - descriptastorus
+  - typed-argument-parser>=1.6.1
```
TAP is not used in the current v2

KnathanM commented on pyproject.toml:
```diff
@@ -5,7 +5,7 @@ build-backend = "setuptools.build_meta"
 name = "chemprop"
 description = "Molecular Property Prediction with Message Passing Neural Networks"
-version = "1.5.1"
+version = "2.0.0a0"
```
Is this changing the pyproject.toml in the top level chemprop folder that still has v1? If so maybe we should wait to update this until we move v2 out of its folder and v1 into a folder.

JacksonBurns replied:
It is indeed changing in the top level folder. I had to do it here since the pip install in the CI was expecting to use a file found in the root directory to do the install. I don't know off the top of my head if you are able to pip install a package if the pyrpoject.toml file is in a different directory.

kevingreenman replied:
I think this is fine since it's on the `v2/dev` branch. I don't think anyone will try to install v1 from this branch.

kevingreenman replied:
However, I think we should change the versioning to match the semantic versioning [specification](https://semver.org/) for pre-releases.

kevingreenman replied:
```suggestion
version = "2.0.0-alpha"
```

kevingreenman replied:
Based on my re-reading of the `semver` and `bump2version` docs, I think I agree with @davidegraff that we should keep this as `1.5.1` for now.

From the semver docs, I think `2.0.0-alpha` will take precedence over `1.5.1` since it doesn't compare pre-release versions unless major, minor, and patch versions are the same (does this mean someone installing `chemprop` with `pip` will get the alpha version instead of the stable version if they don't specify a version, assuming they have python 3.11?).

So should we only run `bump2version` and make a new PyPI release once `2.0.0` is stable?

JacksonBurns replied:
pip will not install releases marked as "pre-releases" unless the user specifically allows it (with `--pre`), so I think in the case that only a pre-release would satisfy installation requirements (like here, if a user tried to install from 3.12) it would still not install (but I might be wrong). I'll add though that we haven't actually put anything on PyPI (and our automated build action doesn't run on this branch), so this isn't possible regardless.

2.0.0-alpha will be changed by python's `build` package to 2.0.0a0 (see [PEP440](https://peps.python.org/pep-0440/) for Python's supported version specifiers) which is why I had it that way.

I went ahead and changed it since I never use bumpversion personally, though I know it's popular. I think it's less work to just manually edit the single-sourced version in pyproject.toml rather than install a utility to do it.

davidegraff replied:
IMO it's useful for when you have multiple sources of the version info (i.e., docs, package, etc.) that can't be "sourced" from the package info (at least I'm not familiar with a way to dynamically build my documentation such that it uses the package version as its source.) In any case, the `2.0.0a0` tag already exists (a minimally functional version back from Nov. '22), so we can call this working branch `a1`.
```suggestion
version = "2.0.0a1"
```

KnathanM reviewed (commented):
I'd like to get this merged in soon so that we can connect the CLI for data splitting to it. And perhaps as the CLI is connected we will find more changes to this code. I didn't review the tests, so that should be done before this PR is merged.

hwpang commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
+                "Number of folds for cross-validation must be between 2 and len(data), inclusive."
+            )
+
+        random = Random(0)
```
Why is the `seed` set to 0 but not using the param `seed`?

JacksonBurns replied:
This has been removed in later commits.

hwpang commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
+
+    train, val, test = None, None, None
+    if split == "crossval":
+        index_set = args.crossval_index_sets[args.seed]
```
What is the difference between using `args.seed` and the input param `seed`?

hwpang replied:
Also if you are allowing `args` to be `None`, these codes can fail if `args` is `None`.

JacksonBurns replied:
args has also now been removed

hwpang commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
+from typing import Tuple
+import os
+
+from chemprop.data import log_scaffold_stats
+from chemprop.args import TrainArgs
```
These two imports are v1 code. We will have to use the v2 equivalence if there's any.

JacksonBurns replied:
import of trainargs has been removed, log import has been fixed

hwpang commented on chemprop/v2/tests/unit/test_data_utils.py:
```diff
@@ -0,0 +1,98 @@
+        """Testing the random split with an empty set"""
+        train, val, test = split_data(data=self.dataset, sizes=(0.4, 0, 0.6))
+        self.assertEqual(val.smiles(), [])
+
```
We should also add a test where `args` is passed in as an input to make sure any part that uses `args` run as expected

JacksonBurns replied:
args no longer present, resolved

JacksonBurns commented on chemprop/v2/data/utils.py:
```diff
@@ -2,65 +2,253 @@
+    key_molecule_index: int = 0,
+    seed: int = 0,
+    num_folds: int = 1,
+    args: TrainArgs = None,
```
removed args, see previous

JacksonBurns commented:
Thanks for updating the title @kevingreenman

JacksonBurns commented:
@KnathanM @hwpang latest comments addressed, please re-review!

davidegraff reviewed (commented):
another one nitpick is to move the `split` arg to a new `SplitType` enum (inherting from `utils.AutoName`) and use that. I know a bunch of packages use strings for stuff like this, but the rest of chemprop uses enum's. Also, I made some comments about erroneous method access because some properties/methods have been removed in the move from v1 to v2, namely `mols` and `smiles`.

hwpang commented on chemprop/v2/tests/unit/test_data_utils.py:
```diff
@@ -0,0 +1,98 @@
+            ["CO", "CCCO"],
+            ["CN", "CCC"],
+        ]
+        self.dataset = MoleculeDataset([MoleculeDatapoint(s) for s in smiles_list])
```
This should be `MoleculeDatapoint.from_smi(s)`

hwpang commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,168 @@
+    return MoleculeDataset(train), MoleculeDataset(val), MoleculeDataset(test)
+
+
+def _unpack_astartes_result(data, result, include_val, log_stats=False):
```
Can you add type for the inputs and the return for this function

hwpang commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,168 @@
 from collections import defaultdict
 from typing import Sequence
+import itertools
 
 import numpy as np
-from rdkit.Chem.Scaffolds import MurckoScaffold
 
 
 from chemprop.v2.data.datapoints import MoleculeDatapoint
+from chemprop.v2.data.datasets import MoleculeDataset
+
+from astartes import train_val_test_split, train_test_split
+from astartes.molecules import train_val_test_split_molecules, train_test_split_molecules
+
+import copy
+from logging import Logger
+import pickle
+from random import Random
+from typing import Tuple
+import os
+
+from chemprop.v2.data.utils import log_scaffold_stats
```
Could you combine/organize the imports? e.g., instead of having `from typing import Tuple` and `from typing import Sequence`, combine them as `from typing import Tuple, Sequence`, and group imports from `chemprop.v2` together.

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,168 @@
+    if split in {"cv-no-val", "cv"}:
+        if num_folds <= 1 or num_folds > len(data):
+            raise ValueError(
+                "Number of folds for cross-validation must be between 2 and len(data), inclusive."
```
Can the number of folds be between 2 and `len(data)`? If they are in a list together (e.g. [2,20]), it looks like the number of folds can only be either 2 or `len(data)` and not anything in between.

JacksonBurns replied:
Gone for the middle ground here, let me know your thoughts

```python
if (max_folds := len(data)) > num_folds or num_folds <= 1:
       raise ValueError(f"Number of folds for cross-validation must be between 2 and {max_folds} (length of data) inclusive (got {num_folds}).")

```

davidegraff replied:
I think Jackson's middle ground is the best- doesn't really cost us anything to be clear and descriptive in our error messages.

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,168 @@
+    key_molecule_index: int = 0,
+    seed: int = 0,
+    num_folds: int = 1,
+    logger: Logger = None,
```
Jackson, you can see an example of logging at the module level in cli/main.py

hwpang commented:
Thanks for addressing the comments! I notice that the tests are failing with the error trace below, but I think we can resolve this in another PR since this PR is blocking some CLI work and should be merged in soon.

```
Traceback:
/usr/share/miniconda3/envs/chemprop/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
chemprop/v2/tests/unit/test_data_utils.py:4: in <module>
    from chemprop.v2.data.utils import split_data
chemprop/v2/data/utils.py:[21](https://github.com/chemprop/chemprop/actions/runs/6412087520/job/17408782385?pr=433#step:5:22): in <module>
    from chemprop.v2.data.utils import log_scaffold_stats
E   ImportError: cannot import name 'log_scaffold_stats' from partially initialized module 'chemprop.v2.data.utils' (most likely due to a circular import) (/home/runner/work/chemprop/chemprop/chemprop/v2/data/utils.py)
```

JacksonBurns commented:
Pushed changes for review comments and rebased onto v2/dev, please re-review! Thanks

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+
+DEFAULT_LOGFILE = f"{LOG_DIR}/{NOW}.log"
+LOG_LEVELS = [logging.ERROR, logging.WARNING, logging.INFO, logging.DEBUG]
+SUBCOMMANDS = []
```
Do we need `DEFAULT_LOGFILE`, `LOG_LEVELS`, and `SUBCOMMANDS` in this file? I don't see them used.

davidegraff replied:
yeah think we can remove

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+        result = split_fun(np.arange(len(data)), sampler="random", **astartes_kwargs)
+        train, val, test = _unpack_astartes_result(data, result, include_val)
+
+    elif split == SplitType.KMEANS:
```
In the README and SplitType class, Kennard Stone comes before k-means, but here in utils and in the tests, the order is flipped. The rest of the split types are consistent in their order in all four places. Not a big deal and not a required change.

JacksonBurns replied:
Good call, went ahead and did it

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+from chemprop.v2.cli.utils import LOG_DIR, NOW
 from chemprop.v2.data.datapoints import MoleculeDatapoint
+from chemprop.v2.data.datasets import MoleculeDataset
+from chemprop.v2.data.utils import log_scaffold_stats
```
This is importing from the file it is in (v2/data/utils.py). There is not `log_scaffold_stats` in this file. From what I can tell, in v1 there was data/scaffold.py which had this function. It should be ported over, perhaps even into astartes, if it is an important part of doing scaffold splits. If it is ported over into chemprop v2 (and not into astartes) we will want the function to not require a logger as it did in v1, but instead to do logging at the module level (as described in another comment).

JacksonBurns replied:
All this function did in v1 was print how many scaffolds were found, as well as the number per scaffold, and this comment reminded me that `astartes` actually does that already. Removing.

KnathanM replied:
Where does astartes print the scaffold information to? If possible we would like to send it to the logging file instead of std.out

JacksonBurns replied:
I should have looked before saying but `astartes` actually only prints out some limited warning information in the case that scaffolds aren't found for a molecule.

JacksonBurns replied:
A question, are we redirecting log output for all our external calls? I think we could do that globally, if so.

kevingreenman commented on .github/workflows/tests.yml:
```diff
@@ -7,51 +7,40 @@ on:
+      shell: bash -l {0}
+      run: |
+        pytest -v chemprop/v2/tests
+    - name: Lint with flake8  #move this back before the tests after v2 development is done
```
```suggestion
    - name: Lint with flake8  # TODO : move this back before the tests after v2 development is done
```

JacksonBurns replied:
Thanks for catching this, going to move to line above and add TODO

kevingreenman commented on .github/workflows/tests.yml:
```diff
@@ -7,51 +7,40 @@ on:
     - uses: actions/checkout@v2
     - uses: conda-incubator/setup-miniconda@v2
       with:
-        python-version: ${{ matrix.python-version }}
```
Is the reason for removing this line just that we're only testing on version 3.11 now? Does it hurt to leave it and also test on 3.12?

JacksonBurns replied:
This was removed because at the time 3.12 was not out, adding it back now (and python 3.12)

kevingreenman commented on README.md:
```diff
@@ -185,6 +185,8 @@ Our code supports several methods of splitting data into train, validation, and
 * **Scaffold.** Alternatively, the data can be split by molecular scaffold so that the same scaffold never appears in more than one split. This can be specified by adding `--split_type scaffold_balanced`.
 * **Random With Repeated SMILES.** Some datasets have multiple entries with the same SMILES. To constrain splitting so the repeated SMILES are in the same split, use the argument `--split_type random_with_repeated_smiles`.
 * **Separate val/test.** If you have separate data files you would like to use as the validation or test set, you can specify them with `--separate_val_path <val_path>` and/or `--separate_test_path <test_path>`. If both are provided, then the data specified by `--data_path` is used entirely as the training data. If only one separate path is provided, the `--data_path` data is split between train  ...
+* **Kennard-Stone** Strictly enforces model interpolation during training to better asses capacity for inference. Useful for high-dimensional and/or small datasets, this algorithm attempts to sample feature space uniformly into the training set. Usage: `--split_type kennard_stone`
```
```suggestion
* **Kennard-Stone** Strictly enforces model interpolation during training to better assess capacity for inference. Useful for high-dimensional and/or small datasets, this algorithm attempts to sample feature space uniformly into the training set. Usage: `--split_type kennard_stone`
```

JacksonBurns replied:
Haha thanks

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+
+    elif split == SplitType.KMEANS:
+        result = mol_split_fun(
+            np.array([m[key_molecule_index] for m in data.smiles()]),
```
not really sure what this is trying to accomplish, but `data.smiles()` does not exist as a method.

davidegraff replied:
you can get a SMILES string via `Chem.MolFromSmiles(data[i].mol)`

KnathanM replied:
@oscarwumit explained to me today that the `key_molecule_index` lets you specify which molecule in your input you want to use for splitting. e.g. ['CO','c1ccccc1'] might represent solute/solvent and you set `key_molecule_index=0` to split by solute. I'm in favor of keeping `key_molecule_index` here.

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+
+    elif split == SplitType.KENNARD_STONE:
+        result = mol_split_fun(
+            np.array([m[key_molecule_index] for m in data.smiles()]),
```
see above

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+    else:
+        train_idxs, test_idxs = result[2], result[3]
+    if log_stats:
+        log_scaffold_stats(data, [set(train_idxs), set(val_idxs), set(test_idxs)])
```
where is this function defined?

JacksonBurns replied:
Removed this following a comment from Nathan above

davidegraff commented on pyproject.toml:
```diff
@@ -56,9 +57,6 @@ demo = "http://chemprop.csail.mit.edu/"
 [project.scripts]
 chemprop = "chemprop.v2.cli.main:main"
 
-[tool.setuptools_scm]
-write_to = "chemprop/_version.py"
-
```
why is this deleted?

JacksonBurns replied:
I just prefer the version to be in the `toml` and there only but we could add this back if there is a consensus. I like the approach I have since it seems more in the spirit of the `toml` file.

davidegraff replied:
This is only for local builds, so if you’re installing from source on a new commit then you can have your version reflect that (I.e., “2.0.1+<git_hash>”

JacksonBurns replied:
Oh cool, adding back

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+    Splits data into training, validation, and test splits.
+
+    :param data: A sequence of :class:`~chemprop.data.MoleculeDatapoint`.
+    :param split_type: Split type.
```
This should be "split" instead of "split_type".

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
+    Splits data into training, validation, and test splits.
+
+    :param data: A sequence of :class:`~chemprop.data.MoleculeDatapoint`.
+    :param split_type: Split type.
```
Wondering if we can have the docstring show what split types are available, something like:
`r"""
    Splits data into training, validation, and test splits.
    :param data: A sequence of :class:`~chemprop.data.MoleculeDatapoint`.
    :param split_type: Split type. Choose one of '''
+ str(SplitTypes.keys())
+ r"""
    :param sizes: A length-3 tuple with the proportions of data in the train, validation, and test sets.
    :param key_molecule_index: For data with multiple molecules, this sets which molecule will be considered during splitting.
    :param seed: The random seed to use before shuffling data.
    :param num_folds: Number of folds to create (only needed for "cv" split type).
    :return: A tuple of :class:`~chemprop.data.MoleculeDataset`\ s containing the train,
             validation, and test splits of the data.
    """`

JacksonBurns replied:
This is a cool suggestion but from what I can see on StackOverflow ([here](https://stackoverflow.com/questions/2693883/dynamic-function-docstring)) it's not really possible. I have added reference to the SplitTypes enum in the docstring.

oscarwumit commented on environment.yml:
```diff
@@ -19,6 +20,4 @@ dependencies:
-  - pip:
-    - typed-argument-parser>=1.6.1
-    - git+https://github.com/bp-kelley/descriptastorus
+  - descriptastorus
```
Would you please add a comment about why "astartes" is not in the environment.yml file? I see that this dependency is provided in the pyproject.toml. However, there are some overlaps between these two files. We probably wanna clean up the dependencies and make then consistent. Thanks.

JacksonBurns replied:
I have updated the environment file to match the pyproject. The only difference is `astartes[molecules]` is now `astartes` and `aimsim` (which `astartes` uses to deal with `molecules`) between the two.

KnathanM commented on chemprop/v2/tests/unit/test_data_utils.py:
Perhaps this file could be moved to `v2/tests/unit/data/test_data_split.py` for organization. I feel it should at least go in a folder, either data or utils.

kevingreenman commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,173 @@
-    k: int = 5,
-    fold: int = 0,
-):
+    split: str | SplitType,
```
why was the default of `random` removed?

JacksonBurns replied:
on accident, adding back

JacksonBurns commented:
Latest comments addressed, please re-review! Thanks

kevingreenman commented on environment.yml:
```diff
@@ -5,20 +5,13 @@ channels:
-  - pip:
-    - typed-argument-parser>=1.6.1
-    - git+https://github.com/bp-kelley/descriptastorus
+  - python>=3.11
```
Was the `pyproject.toml` file missing some depedencies before you copied it over to this file, or are we sure that all of those are no longer required in v2? e.g. I no longer see `hyperopt`, `matplotlib`, `pandas-flavor`, `tensorboardX`, `torchvision`, or `tqdm`. I would think we will at least need to retain `hyperopt` out of this list, although HPO has not yet been implemented for v2.

kevingreenman replied:
Relatedly, is it okay that the minimum versions that were originally specified in this file have been removed for most packages that were retained? And that these minimum versions are also missing from the `pyproject.toml`?

KnathanM replied:
I checked and none of these removed dependencies are used explicitly in v2 as it now stands. I also am in favor of retaining hyperopt as that will be added in the not too distant future (though maybe it would make more sense to add it back as part of that PR). `pandas-flavor` and `torchvision` were not used explicitly in v1 best I can tell, so I don't know if they were/are needed implicitly.

JacksonBurns replied:
The toml and yml were not matched, and I don't know which was 'correct' (if either). As for the specific packages:
 - hyperopt: yes, will need to be added back if we re-implement hpo with hyperopt. Optuna may be a better option, since it plays well with Hydra (as does PL).
 - matplotlib was only used by the tsne script (in scripts), which has not been ported to v2. I also don't think it makes sense to package all of chemprop with matplotlib if only one script (that most people wont use) requires it
 - pandas-flavor is not used anywhere in our code, and if its a dependency of some of our other dependents its up to them to require it
See Nathan's comments about the other deps - this is all I had to add on.

For the package versions, since we moved to Python 3.11+ only versions of those dependencies which satisfy those requirements are eligible anyway.

davidegraff replied:
the `.yml` is way out of date- I think if anything it's the current `v1.6` env file, so I would ignore for `v2` purposes. If we were professionals, we would've all been using the same `.yml` for development purposes, but alas! I think we should probably raise this in another PR though and let it slide for now (i.e., not worry the `.yml` file)

JacksonBurns commented:
I have made the latest updates as per the comments. We need to have an offline discussion about `key_molecule_index`.

davidegraff commented on chemprop/v2/tests/unit/data/test_data_utils.py:
does these tests work? I assume it does, but `TestSplitData` doesn't inherit from anything so how do the `self.assert*` statements do anything?

JacksonBurns replied:
has since been updated and they work now

JacksonBurns commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
+
+        case SplitType.KENNARD_STONE:
+            result = mol_split_fun(
+                np.array([d.mol for d in data]),
```
`astartes` operates on numpy arrays (since the splitting backends actually operate on the values, rather than just the random number generator) and raises warnings about conversions for other data types.

JacksonBurns commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
+        # Use to constrain data with the same smiles go in the same split.
+        case SplitType.RANDOM_WITH_REPEATED_SMILES:
+            # get two arrays: one of all the smiles strings, one of just the unique
+            all_smiles = np.array([d.mol for d in data])
+            unique_smiles = np.unique(all_smiles)
```
nice but not going to take, need to be able to index into unique_smiles

JacksonBurns commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
+                copied_mol = copy.deepcopy(mol)
+                for atom in copied_mol.GetAtoms():
+                    atom.SetAtomMapNum(0)
+                mols_without_atommaps.append([copied_mol])
```
yes unforunately `astartes` expects 2d lists

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -53,42 +52,42 @@ def split_data(
 
     train, val, test = None, None, None
-    match split:
+    match SplitType(split):
```
```suggestion
    match SplitType.get(split):
```

davidegraff reviewed (commented):
just the one change and I think this is good. Haven't tested myself, but I like playing with fire and assuming it works

JacksonBurns commented:
Tests are now passing for `split_data`: https://github.com/chemprop/chemprop/actions/runs/6433655460/job/17471164794?pr=433#step:5:17

More reviews, please.

davidegraff reviewed (commented):
sorry I caught one more thing. Just needs to merge `v2/dev` and we're done. @hwpang all good?

KnathanM commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
+
+def _unpack_astartes_result(
+    data: Sequence[MoleculeDatapoint], result: tuple, include_val: bool
+) -> Tuple[list[MoleculeDatapoint], list[MoleculeDatapoint], list[MoleculeDatapoint]]:
```
This still has a capital `Tuple`. It can be lowercased and then import Tuple removed. I'm also happy to just merge the PR and not worry about this.

hwpang reviewed (approved):
Looks good to me.

JacksonBurns commented:
Final changes addressed, rebased onto v2/dev. Should be good to go! Thanks for you work on this, everyone.

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
+) -> tuple[list[MoleculeDatapoint], list[MoleculeDatapoint], list[MoleculeDatapoint]]:
+    """Helper function to partition input data based on output of astartes sampler
+
+    Args:
+        data (MoleculeDataset): The data being partitioned. If None, returns indices.
+        result (tuple): Output from call to astartes containing the split indices
+        include_val (bool): True if a validation set is included, False otherwise.
+
+    Returns:
+        MoleculeDatset: The train, validation (can be empty) and test dataset
```
```suggestion
    Parameters
    -----------
    data: MoleculeDataset
        The data being partitioned. If None, returns indices.
    result: tuple
        Output from call to astartes containing the split indices
    include_val: bool
        True if a validation set is included, False otherwise.

    Returns
    ---------
    train: MoleculeDataset
    val: MoleculeDataset
        NOTE: possibly empty
    test: MoleculeDataset
```

davidegraff commented on chemprop/v2/data/utils.py:
```diff
@@ -1,66 +1,163 @@
-):
+    seed: int = 0,
+    num_folds: int = 1,
+) -> tuple[Sequence[MoleculeDatapoint], ...]:
+    r"""
+    Splits data into training, validation, and test splits.
+
+    :param datapoints: A sequence of :class:`~chemprop.data.MoleculeDatapoint`.
+    :param split: Split type, one of ~chemprop.data.utils.SplitType
+    :param sizes: A length-3 tuple with the proportions of data in the train, validation, and test sets.
+    :param seed: The random seed to use before shuffling data.
+    :param num_folds: Number of folds to create (only needed for "cv" split type).
+    :return: A tuple of Sequences of :class:`~chemprop.data.MoleculeDatapoint`\ s containing the train,
+             validation, and test splits of the data.
+    """
```
switch to numpy docstring format.

davidegraff reviewed (commented):
I somehow keep finding things to nitpick 🤗 . I promise i will stop after this.

JacksonBurns commented:
Rebased with latest updates - should be finally super-duper very ready
